### PR TITLE
@damassi => Refactor Draft Utils

### PR DESCRIPTION
**Pull request by eessex**
Hold pending minor CSS update to Reaction.

A long-needed close look at our text editor post Article2.  This PR updates Draft and incorporates some recent changes to the library, allowing us to remove instances of manual text selection and clean up some other selection-related functions. Lack of event fallbacks for R/L arrows meant we were manually setting selections with arrows, causing unexpected bugs and behavior. 

Also kills all the 'createEntity'  warnings on import/export, making Positron ok for React 16+.

The `content-start` entity (for drop-caps) has been removed in favor of plain CSS. 

`content-end` still needs to be inserted into html as a span. This has been refactored to improve the editing experience and stop a bug where it was saving in unexpected places. 

The `rich_text` utils folder has been decaffeinated and broken out into files by purpose.  This brings much better clarity and readability, I plan to follow up soon with a refactor of the text section component in plain js as well.

Related to:
https://github.com/artsy/positron/issues/1414
https://github.com/artsy/publishing/issues/231
https://github.com/artsy/positron/issues/1376
https://github.com/artsy/positron/issues/1260
https://github.com/artsy/positron/issues/1273
https://github.com/artsy/positron/issues/1192

### CODE REVIEW

File: client/apps/edit/components/content/sections/text/index.coffee [hunk 1]
```diff
@@ -7,7 +7,20 @@ Config = require './draft_config.js'
...
+  getSelectionDetails
+} = require '../../../../../../components/rich_text/utils/text_selection.js'
+{ setContentEnd } = require '../../../../../../components/rich_text/utils/decorators.js'
+{
+  standardizeSpacing,
+  stripCharacterStyles,
+  stripGoogleStyles
+} = require '../../../../../../components/rich_text/utils/text_stripping.js'
+{ convertFromRichHtml, convertToRichHtml } = require '../../../../../../components/rich_text/utils/convert_html.js'
+{ keyBindingFnFull } = require '../../../../../../components/rich_text/utils/keybindings.js'
```
Comment (eessex) [inline client/apps/edit/components/content/sections/text/index.coffee]: These imports are long, but will be fixed up in a coming PR refactoring this component.

File: client/apps/edit/components/content/sections/text/index.styl [hunk 2]
```diff
@@ -29,4 +29,11 @@
       &:before
         content url(/icons/circle_plus.svg) !important
 
-
+.feature
```
Comment (eessex) [inline client/apps/edit/components/content/sections/text/index.styl]: Should be removed before merging -- pending Reaction update

**Comment by damassi** [comment]
PR's like this are the best :) One thing I would suggest is if you're using `decaffinate` you should also run it through prettier as a second pass (though I'm not suggesting you hook up Prettier to the project, unless you want to) https://prettier.io/docs/en/cli.html

**Comment by eessex** [comment]
Will look into prettier, was not automated for the most part and gone through with a careful hand if so.  Looks like the test runners are still failing when they hit jest -- not sure if this had to do with webpack changes?

**Comment by damassi** [comment]
Looking at Circle the problems definitely look test-related and the tests were working before: 

screen shot 2017-11-17 at 12 30 03 pm

(extra space)

But yeah, prettier will take like 80% of the work out of converting those files to readable JS.

**Comment by eessex** [comment]
Oh hm true

**Comment by eessex** [comment]
Was seeing that issue in earlier iterations today - https://circleci.com/gh/artsy/positron/971

**Comment by damassi** [comment]
> Was seeing that issue in earlier iterations today

Damn :( I don't have time to debug this right now, but could you maybe post in #platform or #dev and see if anyone has any thoughts? Attached is a CSV showing how the test runner seems to be stuck in a recursive spawn loop: 

[memory-usage.csv.zip](https://github.com/artsy/positron/files/1483838/memory-usage.csv.zip)

**Comment by kanaabe** [comment]
When you're feelin better let's see if pulling from master fixes this one via  https://github.com/artsy/positron/pull/1454

File: client/apps/edit/components/content/sections/text/draft_config.js [hunk 3]
```diff
@@ -1,11 +1,10 @@
 import Immutable from 'immutable'
 import {
-  ContentStartEnd,
+  ContentEnd,
   findContentEndEntities,
-  findContentStartEntities,
   findLinkEntities,
   Link
-} from 'client/components/rich_text/utils/decorators.coffee'
+} from 'client/components/rich_text/utils/decorators.js'
```
Comment (damassi) [inline client/apps/edit/components/content/sections/text/draft_config.js]: Can remove extension

File: client/apps/edit/components/content/sections/text/index.coffee [hunk 4]
```diff
@@ -39,28 +56,28 @@ module.exports = React.createClass
...
     @setState
       html: html
       editorState: editorState
 
   componentDidUpdate: (prevProps) ->
-    if @props.isEndText isnt prevProps.isEndText or @props.isStartText isnt prevProps.isStartText
+    if @props.isContentEnd isnt prevProps.isContentEnd
       unless @props.article.get('layout') is 'classic'
-        html = Utils.setContentStartEnd(@props.section.get('body'), @props.article.get('layout'), @props.isStartText, @props.isEndText)
+        html = setContentEnd(@props.section.get('body'), @props.isContentEnd)
```
Comment (damassi) [inline client/apps/edit/components/content/sections/text/index.coffee]: 👍

File: client/components/rich_text/test/components/paragraph.test.coffee [hunk 5]
```diff
@@ -216,8 +213,7 @@ describe 'Rich Text: Paragraph', ->
       @component.onPaste('hello here again.', '<p>hello</p><p>here again.</p>')
       @stripGoogleStyles.called.should.eql true
 
-    it 'calls standardizeSpacing', ->
-      @Utils.standardizeSpacing = sinon.stub()
+    xit 'calls standardizeSpacing', ->
```
Comment (damassi) [inline client/components/rich_text/test/components/paragraph.test.coffee]: Meant to be x'd out?

File: client/components/rich_text/test/utils/keybindings.test.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
+import {
+  keyBindingFnParagraph,
+  keyBindingFnFull
+} from '../../utils/keybindings.js'
+
+jest.unmock('draft-js')
```
Comment (damassi) [inline client/components/rich_text/test/utils/keybindings.test.js]: Does jest still auto-mock? I thought they got rid of that default?

File: client/components/rich_text/test/components/paragraph.test.coffee (hunk 5 above)
Comment (eessex) [reply client/components/rich_text/test/components/paragraph.test.coffee]: Was having some trouble stubbing this one -- but will to come back to it next week when the actual components are decaffeinated.

**Pull request merged**